### PR TITLE
[Remove] Multiple Types from IndexTemplateMetadata

### LINKED ISSUE #2359: [BUG] org.opensearch.upgrades.FullClusterRestartIT > testEmptyShard FAILED
**Describe the bug**
```
REPRODUCE WITH: ./gradlew ':qa:full-cluster-restart:v1.3.0#oldClusterTest' --tests "org.opensearch.upgrades.FullClusterRestartIT.testEmptyShard" -Dtests.seed=3DE5D0AB12901A69 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=ar-AE -Dtests.timezone=America/Dawson -Druntime.java=17

```

````
org.opensearch.upgrades.FullClusterRestartIT > testEmptyShard FAILED
    org.opensearch.client.WarningFailureException: method [PUT], host [http://[::1]:37365], URI [/test_empty_shard], status line [HTTP/1.1 200 OK]
    {"acknowledged":true,"shards_acknowledged":true,"index":"test_empty_shard"}
        at __randomizedtesting.SeedInfo.seed([3DE5D0AB12901A69:719E0773EEA935FC]:0)
        at app//org.opensearch.client.RestClient.convertResponse(RestClient.java:346)
        at app//org.opensearch.client.RestClient.performRequest(RestClient.java:320)
        at app//org.opensearch.client.RestClient.performRequest(RestClient.java:295)
        at app//org.opensearch.test.rest.OpenSearchRestTestCase.createIndex(OpenSearchRestTestCase.java:976)
        at app//org.opensearch.test.rest.OpenSearchRestTestCase.createIndex(OpenSearchRestTestCase.java:956)
        at app//org.opensearch.test.rest.OpenSearchRestTestCase.createIndex(OpenSearchRestTestCase.java:952)
        at app//org.opensearch.upgrades.FullClusterRestartIT.testEmptyShard(FullClusterRestartIT.java:670)
````

     

**Host/Environment (please complete the following information):**

**Additional context**

### PR DESCRIPTION
**Pull request by nknize**
Removes multi-type support from IndexTemplateMetadata so that instead of holding
a map of multiple types to mappings, it only returns a single mapping for a
single type. Also removes type from documentMapper() method to avoid any 
accidental NullPointerExceptions in the internal mapping retrieval.

relates #1940
closes #2359

### CODE REVIEW

**Comment by nknize** [comment]
Will cleanup after #2397 merges...

File: client/rest-high-level/src/test/java/org/opensearch/client/indices/GetIndexTemplatesResponseTests.java [hunk 1]
```diff
@@ -224,7 +223,10 @@ static void toXContent(GetIndexTemplatesResponse response, XContentBuilder build
             serverTemplateBuilder.order(clientITMD.order());
             serverTemplateBuilder.version(clientITMD.version());
             if (clientITMD.mappings() != null) {
-                serverTemplateBuilder.putMapping(MapperService.SINGLE_MAPPING_NAME, clientITMD.mappings().source());
+                // The client-side mappings never include a wrapping type, but server-side mappings
+                // for index templates still do so we need to wrap things here
+                String mappings = "{\"" + MapperService.SINGLE_MAPPING_NAME + "\": " + clientITMD.mappings().source().string() + "}";
```
Comment (reta) [inline client/rest-high-level/src/test/java/org/opensearch/client/indices/GetIndexTemplatesResponseTests.java:228]: :+1:  but will be fixed later?

**Comment by dreamer-89** [comment]
`distribution:bwc:minor:buildBwcLinuxTar` failure. 

```
FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':distribution:bwc:minor:buildBwcLinuxTar'.
> Building 1.3.0 didn't generate expected file /var/CITOOL/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/search/distribution/bwc/minor/build/bwc/checkout-1.x/distribution/archives/linux-tar/build/distributions/opensearch-min-1.3.0-SNAPSHOT-linux-x64.tar.gz

```

**Comment by dreamer-89** [comment]
start gradle check

**Comment by dreamer-89** [comment]
> > `distribution:bwc:minor:buildBwcLinuxTar` failure. Getting fixed in #2430
> > ```
> > FAILURE: Build failed with an exception.
> > 
> > * What went wrong:
> > Execution failed for task ':distribution:bwc:minor:buildBwcLinuxTar'.
> > > Building 1.3.0 didn't generate expected file /var/CITOOL/workflow/OpenSearch_CI/PR_Checks/Gradle_Check/search/distribution/bwc/minor/build/bwc/checkout-1.x/distribution/archives/linux-tar/build/distributions/opensearch-min-1.3.0-SNAPSHOT-linux-x64.tar.gz
> > ```
> 
> @dreamer-89 The pull request was merged moments ago, may need rebase

@nknize : Can you please rebase against main ?

**Review by dreamer-89 — APPROVED** [review]
LGTM!

Thank you for taking time to raise this PR.

File: client/rest-high-level/src/test/java/org/opensearch/client/indices/GetIndexTemplatesResponseTests.java (hunk 1 above)
Comment (nknize) [reply client/rest-high-level/src/test/java/org/opensearch/client/indices/GetIndexTemplatesResponseTests.java:228]: Yes

File: server/src/main/java/org/opensearch/action/admin/indices/create/CreateIndexRequest.java [hunk 2]
```diff
@@ -278,18 +294,17 @@ public CreateIndexRequest mapping(String type, XContentBuilder source) {
...
-        if (mappings.containsKey(type)) {
-            throw new IllegalStateException("mappings for type \"" + type + "\" were already defined");
-        }
         // wrap it in a type map if its not
         if (source.size() != 1 || !source.containsKey(type)) {
-            source = MapBuilder.<String, Object>newMapBuilder().put(type, source).map();
+            source = Collections.singletonMap(MapperService.SINGLE_MAPPING_NAME, source);
+        } else if (MapperService.SINGLE_MAPPING_NAME.equals(type) == false) {
+            // if it has a different type name, then unwrap and rewrap with _doc
+            source = Collections.singletonMap(MapperService.SINGLE_MAPPING_NAME, source.get(type));
```
[the remark this answers is not in the document]
Comment (nknize) [reply server/src/main/java/org/opensearch/action/admin/indices/create/CreateIndexRequest.java:302]: For template bwc; same as above.

**Comment by dreamer-89** [comment]
```
> Task :qa:full-cluster-restart:v1.2.5#oldClusterTest

REPRODUCE WITH: ./gradlew ':qa:full-cluster-restart:v1.2.5#oldClusterTest' --tests "org.opensearch.upgrades.FullClusterRestartIT.testEmptyShard" -Dtests.seed=B30D22FAABA64229 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=es-BO -Dtests.timezone=Brazil/DeNoronha -Druntime.java=17

org.opensearch.upgrades.FullClusterRestartIT > testEmptyShard FAILED
    org.opensearch.client.WarningFailureException: method [PUT], host [http://[::1]:35241], URI [/test_empty_shard], status line [HTTP/1.1 200 OK]
    Warnings: [Translog retention settings [index.translog.retention.age] and [index.translog.retention.size] are deprecated and effectively ignored. They will be removed in a future version.]
    {"acknowledged":true,"shards_acknowledged":true,"index":"test_empty_shard"}
        at __randomizedtesting.SeedInfo.seed([B30D22FAABA64229:FF76F522579F6DBC]:0)
        at app//org.opensearch.client.RestClient.convertResponse(RestClient.java:346)
        at app//org.opensearch.client.RestClient.performRequest(RestClient.java:320)
        at app//org.opensearch.client.RestClient.performRequest(RestClient.java:295)
        at app//org.opensearch.test.rest.OpenSearchRestTestCase.createIndex(OpenSearchRestTestCase.java:973)
        at app//org.opensearch.test.rest.OpenSearchRestTestCase.createIndex(OpenSearchRestTestCase.java:956)
        at app//org.opensearch.test.rest.OpenSearchRestTestCase.createIndex(OpenSearchRestTestCase.java:952)
        at app//org.opensearch.upgrades.FullClusterRestartIT.testEmptyShard(FullClusterRestartIT.java:670)
```

**Comment by dreamer-89** [comment]
start gradle check

**Comment by nknize** [comment]
> This issue is getting tracked in #2359



See https://github.com/opensearch-project/OpenSearch/issues/2359#issuecomment-1064710886

**Comment by nknize** [comment]
Another #1746; refiring

```
./gradlew ':server:internalClusterTest' --tests "org.opensearch.gateway.RecoveryFromGatewayIT.testReuseInFileBasedPeerRecovery" -Dtests.seed=488FED743787DA93 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=sr-Latn-ME -Dtests.timezone=America/Recife -Druntime.java=17
```

```
org.opensearch.gateway.RecoveryFromGatewayIT > testReuseInFileBasedPeerRecovery FAILED
    java.lang.AssertionError: shard [test][0] on node [node_t1] has pending operations:
     --> RetentionLeaseBackgroundSyncAction.Request{retentionLeases=RetentionLeases{primaryTerm=1, version=1591, leases={peer_recovery/KE4lkT4_Sk60Kzfrtc_G6w=RetentionLease{id='peer_recovery/KE4lkT4_Sk60Kzfrtc_G6w', retainingSequenceNumber=1646, timestamp=1646966789226, source='peer recovery'}, peer_recovery/VA3o7X03RpmgTx04G6DugA=RetentionLease{id='peer_recovery/VA3o7X03RpmgTx04G6DugA', retainingSequenceNumber=1646, timestamp=1646966789226, source='peer recovery'}}}, shardId=[test][0], timeout=1m, index='test', waitForActiveShards=0}
    	at org.opensearch.index.shard.IndexShardOperationPermits.acquire(IndexShardOperationPermits.java:248)
    	at org.opensearch.index.shard.IndexShard.acquirePrimaryOperationPermit(IndexShard.java:3208)
    	at org.opensearch.action.support.replication.TransportReplicationAction.acquirePrimaryOperationPermit(TransportReplicationAction.java:1116)
    	at org.opensearch.action.support.replication.TransportReplicationAction$AsyncPrimaryAction.doRun(TransportReplicationAction.java:433)
    	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:50)
    	at org.opensearch.action.support.replication.TransportReplicationAction.handlePrimaryRequest(TransportReplicationAction.java:377)
    	at org.opensearch.transport.RequestHandlerRegistry.processMessageReceived(RequestHandlerRegistry.java:98)
    	at org.opensearch.transport.TransportService$8.doRun(TransportService.java:937)
    	at org.opensearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:792)
    	at org.opensearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:50)
    	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
    	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
```

**Comment by nknize** [comment]
start gradle check

**Comment by nknize** [comment]
Another #1561 failure; refiring

```
/gradlew ':server:internalClusterTest' --tests "org.opensearch.cluster.allocation.ClusterRerouteIT.testDelayWithALargeAmountOfShards" -Dtests.seed=7F5EA091AFD9EF57 -Dtests.security.manager=true -Dtests.jvm.argline="-XX:TieredStopAtLevel=1 -XX:ReservedCodeCacheSize=64m" -Dtests.locale=vi -Dtests.timezone=EET -Druntime.java=17
```

```
org.opensearch.cluster.allocation.ClusterRerouteIT > testDelayWithALargeAmountOfShards FAILED
    java.lang.AssertionError: AcknowledgedResponse failed - not acked
    Expected: <true>
         but: was <false>
        at __randomizedtesting.SeedInfo.seed([7F5EA091AFD9EF57:532FAA70BEC7C5B5]:0)
        at org.hamcrest.MatcherAssert.assertThat(MatcherAssert.java:18)
        at org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked(OpenSearchAssertions.java:128)
        at org.opensearch.test.hamcrest.OpenSearchAssertions.assertAcked(OpenSearchAssertions.java:132)
        at org.opensearch.test.TestCluster.wipeIndices(TestCluster.java:167)
        at org.opensearch.test.TestCluster.wipe(TestCluster.java:90)
        at org.opensearch.test.OpenSearchIntegTestCase.afterInternal(OpenSearchIntegTestCase.java:601)
        at org.opensearch.test.OpenSearchIntegTestCase.cleanUpCluster(OpenSearchIntegTestCase.java:2235)
```

**Comment by nknize** [comment]
start gradle check

**Pull request merged**